compiler/rustc_errors/src/diagnostic.rs

```diff
@@ -1064,3 +1052,39 @@ impl PartialEq for Diagnostic {
         self.keys() == other.keys()
     }
 }
+
+pub enum HelpUseLatestEdition {
+    Cargo,
+    Standalone,
+}
+
+impl HelpUseLatestEdition {
+    pub fn new() -> Self {
+        if std::env::var_os("CARGO").is_some() { Self::Cargo } else { Self::Standalone }
+    }
+}
+
+impl AddToDiagnostic for HelpUseLatestEdition {
+    fn add_to_diagnostic_with<F>(self, diag: &mut Diagnostic, f: F)
+    where
+        F: Fn(&mut Diagnostic, SubdiagnosticMessage) -> SubdiagnosticMessage,
+    {
+        let msg = f(
+            diag,
+            match self {
+                Self::Cargo => {
+                    format!("set `edition = \"{}\"` in `Cargo.toml`", LATEST_STABLE_EDITION)
+                }
+                Self::Standalone => {
+                    format!("pass `--edition {}` to `rustc`", LATEST_STABLE_EDITION)
+                }
+            }
+            .into(),
+        );
+        diag.help(msg);
+
+        let msg =
+            f(diag, "for more on editions, read https://doc.rust-lang.org/edition-guide".into());
+        diag.note(msg);
```

Comment:
I want to come back to this; this impl is not translatable, which will be a problem in the future. 

IMO there are two ways to deal with that:
- duplicate this diagnostic in both `rustc_hir_typeck` and `rustc_parse` so the derive macro can be used. This also avoids having this in `rustc_errors`.
- move the strings to `errors.ftl` and use `diag.set_arg` and friends rather than manual formatting.

Otherwise I am happy with this PR  🙂

Reply:
I went with option 1, the duplication isn't great but it seemed cleaner overall.